src/compiler/checker.ts

```diff
@@ -6890,11 +6890,11 @@ namespace ts {
                             let s = signature;
                             // Union the result types when more than one signature matches
                             if (unionSignatures.length > 1) {
-                                let thisParameter = signature.thisParameter;
-                                if (forEach(unionSignatures, sig => sig.thisParameter)) {
+                                let thisParameter = forEach(unionSignatures, sig => sig.thisParameter);
+                                if (thisParameter) {
                                     // TODO: GH#18217 We tested that *some* has thisParameter and now act as if *all* do
                                     const thisType = getUnionType(map(unionSignatures, sig => sig.thisParameter ? getTypeOfSymbol(sig.thisParameter) : anyType), UnionReduction.Subtype);
-                                    thisParameter = createSymbolWithType(signature.thisParameter!, thisType);
+                                    thisParameter = createSymbolWithType(thisParameter, thisType);

```

Comment:
We definitely need test for this. The best way to get that would be to run tsc in debug mode without fix and figure out which type in your code base was resulting in the crash and get those signatures to make the union type.
Note that seems like your want union of signatures in which first signature doesn't have this parameter.